### PR TITLE
2754 partial instead of error

### LINKED ISSUE #2754: Return partial dataset-hub-cache instead of error?
`dataset-hub-cache` depends on multiple previous steps, and any error in one of them makes it fail. It provokes things like https://github.com/huggingface/moon-landing/issues/9799 (internal): in the datasets list, a dataset is not marked as "supporting the dataset viewer", whereas the only issue is that we didn't manage to list the compatible libraries, to create the tags.

https://github.com/huggingface/dataset-viewer/blob/main/services/worker/src/worker/job_runners/dataset/hub_cache.py

In this case, we could return a partial response, or maybe return an empty list of libraries or modalities if we have an error.

What do you think @lhoestq?

### PR DESCRIPTION
**Pull request by severo**
fix #2754

### CODE REVIEW

**Comment by severo** [comment]
I don't increment the version, let's wait for https://github.com/huggingface/dataset-viewer/issues/2898, to save some resources.

File: services/worker/src/worker/job_runners/dataset/hub_cache.py [hunk 1]
```diff
@@ -51,23 +51,29 @@ def compute_hub_cache_response(dataset: str) -> tuple[DatasetHubCacheResponse, f
...
+                "Previous step 'dataset-size' did not return the expected content: 'partial' or 'size.dataset.num_rows'."
+            )
+        partial = content["partial"]
+        num_rows = content["size"]["dataset"]["num_rows"]
+        size_progress = size_response["progress"]
+    except (CachedArtifactNotFoundError, PreviousStepFormatError):
+        raise
+    except Exception:
+        partial = False
+        num_rows = 0
```
Comment (lhoestq) [inline services/worker/src/worker/job_runners/dataset/hub_cache.py:75]: the hub won't show 0 rows on the dataset page right ?

**Review by lhoestq — APPROVED** [review]
LGTM !

File: services/worker/src/worker/job_runners/dataset/hub_cache.py (hunk 1 above)
Comment (severo) [reply services/worker/src/worker/job_runners/dataset/hub_cache.py:75]: hmmm maybe it should be optional... Looking at it

Comment (severo) [reply services/worker/src/worker/job_runners/dataset/hub_cache.py:75]: Possible changes soon to this response:
- make num_rows an optional field?
- add num_rows_estimated (see https://github.com/huggingface/dataset-viewer/pull/2906)
- remove "tags"?
- replace "viewer / preview / partial" with one field that would simplify the logic on the Hub side (currently: `!hub_cache.viewer ? "preview" : hub_cache.partial ? "viewer-partial" : "viewer"`) -> `viewer: "viewer" | "viewer-partial" | "preview" | "none"`?

We might as well work all this in a new step (`dataset-moonlanding-cache` for example). And once we've finished this batch of changes, prepare the Hub to consume the new step, and remove `dataset-hub-cache` here.

wdyt @lhoestq?

Comment (severo) [reply services/worker/src/worker/job_runners/dataset/hub_cache.py:75]: It's currently not used on moonlanding btw. So, we can change it. I agree that `num_rows = 0` is wrong.

Two options:
- `num_rows = None`
- don't set the `num_rows` field

Let's set it to `None`

File: services/worker/src/worker/dtos.py [hunk 2]
```diff
@@ -344,7 +344,7 @@ class DatasetHubCacheResponse(TypedDict):
     preview: bool
     viewer: bool
     partial: bool
-    num_rows: int
+    num_rows: Union[int, None]
```
Comment (lhoestq) [inline services/worker/src/worker/dtos.py:347]: Optional in python means Union[int, None]

```suggestion
    num_rows: Optional[int]
```

**Review by lhoestq — APPROVED** [review]

File: services/worker/src/worker/dtos.py (hunk 2 above)
Comment (severo) [reply services/worker/src/worker/dtos.py:347]: sure

**Comment by severo** [comment]
Once deployed, I'll relaunch the datasets with error

**Pull request merged**